feat: RUN-975: Add WebAssembly SIMD example

dfinity-berestovskyy opened the pull request:
**Overview**
This example showcases different approaches to utilizing the new SIMD instructions: Rust auto-vectorization and SIMD intrinsics for matrix multiplication, a core operation in Machine Learning and Artificial Intelligence applications. The example compares various SIMD optimization techniques and their potential speedups.

**Requirements**
To run the example, the latest `dfx` version `0.20.2-beta.0` or later is required with WebAssembly SIMD enabled replica.

ulan commented on rust/simd/mat_mat_mul.wat:
Was this file added by accident?

ulan commented on rust/simd/mat_mat_mul/src/lib.rs:
```diff
@@ -0,0 +1,246 @@
+    }
+    // Assert the generic and SIMD matrix multiplication results are the same.
+    assert_eq!(
+        add_mat_mul_4x4_generic::<f32>(K, pa, pb),
```
Wdyt about additionally having the simplest possible "school" multiplication of matrices as the expected value for assertion? It is not obvious to me that add_mat_mul_4x4_generic correctly computes matrix multiplication because it uses the same 4x4 tiling as other algorithms. If there is a bug in the tiling, we wouldn't catch it.

By simplest matrix multiplication I mean: r[i][j] = sum(a[i][k] * b[k][j]) for all i, j, k.

ulan commented on rust/simd/mat_mat_mul/src/lib.rs:
```diff
@@ -0,0 +1,246 @@
+#[ic_cdk_macros::query]
+fn auto_vectorization_f32() -> u64 {
+    let a = mats::M_4X4_F32.repeat(K / 4);
+    let b = mats::M_4X4_F32.repeat(K / 4);
```
I was expecting that the number of rows in matrix A matches the number of columns in matrix B in order for multiplication to work.

It seems that we are multiplying matrix A with a transposed matrix B (rows and columns swapped?). If so, then let's add a comment about it to avoid confusing readers?

ulan commented on rust/simd/mat_mat_mul/src/lib.rs:
```diff
@@ -0,0 +1,246 @@
+}
+
+/// Returns the number of instructions used for a loop performing
+/// element-wise multiplication of `224x4` tiles of matrices `A` and `B`
```
Nit: looks like we are multiplying a 4x224 (4 rows and 224 columns) matrix A with a 224x4 matrix B to get a 4x4 resulting matrix? Initially I thought that the resulting matrix will be 224x224 if matrix A is 224x4.

ulan commented on rust/simd/README.md:
```diff
@@ -0,0 +1,215 @@
+
+The `mat_mat_mul` canister provide the following interface:
+
+- `generic_f32` &mdash; returns the number of instructions used for a loop performing element-wise multiplication of `224x4` tiles of matrices `A` and `B` using generic Rust operators.
```
See my comments below about the 224x4 dimensions.

ulan commented on rust/simd/Makefile:
```diff
@@ -0,0 +1,44 @@
+.PHONY: install
+.SILENT: install
+install: build
+	dfx canister install heartbeat --argument 1
```
I don't understand the heartbeat and timer parts here. Is this a copy from another project?

ulan commented on rust/simd/README.md:
```diff
@@ -0,0 +1,215 @@
+using just Rust's loop auto-vectorization.
+
+As shown in Example 2, integer operations also benefit, although with a more modest
+"2x" speedup.
```
I think we can get higher speedup for integer using 8-bit integers and doing 16x16 tiles. Just fyi, nothing actionable here.

ulan reviewed (approved):
LGTM with comments. Thanks!

berestovskyy committed:
feat: RUN-975: Add WebAssembly SIMD example

dfx-json commented:
@dfinity-berestovskyy this example should also have a corresponding workflow for automated tests. it can be as simple as just deploying the canister locally.

berestovskyy committed:
Post-review fixes

berestovskyy committed:
Update CODEOWNERS and add a test

dfinity-berestovskyy commented:
Thanks for the review, folks.
I've addressed the comments, updated the CODEOWNERS and added the tests. The tests fail for now, as there is no `dfx` version with SIMD support released yet.

berestovskyy committed:
Make sure there are no compiler optimizations

dfx-json commented:
LGTM. Thank you @dfinity-berestovskyy.

berestovskyy committed:
Fix SIMD multiplication

dfx-json commented:
Pending merge until dfx supports simd and tests pass

berestovskyy committed:
Add global `target-feature` option

berestovskyy committed:
Rename `auto_vectorization` to `auto_vectorized`

berestovskyy committed:
Use `instruction_counter`

berestovskyy committed:
Update `dfx` version

berestovskyy committed:
Update `README`
